File: packages/sdk-core/src/__tests__/dsn.fixtures.test.ts
import { readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

import { describe, expect, it } from 'vitest';

import { InvalidDsnError, parseDsn } from '../dsn.js';

// Single canonical fixture file consumed by the TS, Java and Python SDK test suites. Adding a
// new edge case here means all three SDKs run it on next CI; whichever fails the parity check
// gets fixed. Prevents the "fixed it in TS, forgot Java" drift that bit us 2026-05-09.
//
// Path math: this file lives at packages/sdk-core/src/__tests__/dsn.fixtures.test.ts; the
// fixtures live at scripts/dsn-test-fixtures.json — four levels up plus into scripts.
const FIXTURES_PATH = resolve(
  dirname(fileURLToPath(import.meta.url)),
  '../../../../scripts/dsn-test-fixtures.json',
);

interface Fixture {
  name: string;
  dsn: string;
  valid: boolean;
  publicKey?: string;
  host?: string;
  projectId?: string;
  scheme?: 'http' | 'https';
  ingestUrl?: string;
}

const fixtures = JSON.parse(readFileSync(FIXTURES_PATH, 'utf-8')) as Fixture[];

describe('parseDsn (shared fixtures)', () => {
  for (const fx of fixtures) {
    it(fx.name, () => {
      if (!fx.valid) {
        expect(() => parseDsn(fx.dsn)).toThrow(InvalidDsnError);
        return;
      }
      const out = parseDsn(fx.dsn);
      if (fx.scheme !== undefined) expect(out.protocol).toBe(fx.scheme);
      if (fx.publicKey !== undefined) expect(out.publicKey).toBe(fx.publicKey);
      if (fx.host !== undefined) expect(out.host).toBe(fx.host);
      if (fx.projectId !== undefined) expect(out.projectId).toBe(fx.projectId);
      if (fx.ingestUrl !== undefined) expect(out.ingestUrl).toBe(fx.ingestUrl);
    });
  }
});

File: java-sdk/src/test/java/org/arguslog/sdk/DsnFixturesTest.java
package org.arguslog.sdk;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

/**
 * Cross-SDK parity test. The canonical fixture lives at scripts/dsn-test-fixtures.json (repo root);
 * the TS and Python SDKs run identical assertions against the same file. Adding an edge case there
 * means all three SDKs run it on next CI; whichever fails the parity check gets fixed. Prevents the
 * "fixed it in TS, forgot Java" drift that bit us 2026-05-09.
 *
 * <p>Path math: Gradle sets cwd = projectDir = java-sdk/, so the repo-relative {@code
 * ../scripts/dsn-test-fixtures.json} resolves regardless of where the test runner sits.
 */
class DsnFixturesTest {

  private static final ObjectMapper MAPPER = new ObjectMapper();

  @TestFactory
  List<DynamicTest> sharedFixtures() throws IOException {
    File fixturesFile = new File("../scripts/dsn-test-fixtures.json");
    JsonNode root = MAPPER.readTree(fixturesFile);
    List<DynamicTest> tests = new ArrayList<>();
    for (JsonNode fx : root) {
      String name = fx.get("name").asText();
      tests.add(DynamicTest.dynamicTest(name, () -> assertFixture(fx)));
    }
    return tests;
  }

  private void assertFixture(JsonNode fx) {
    String dsn = fx.get("dsn").asText();
    boolean valid = fx.get("valid").asBoolean();

    if (!valid) {
      assertThatIllegalArgumentException().isThrownBy(() -> Dsn.parse(dsn));
      return;
    }

    Dsn parsed = Dsn.parse(dsn);
    if (fx.has("scheme")) assertThat(parsed.scheme()).isEqualTo(fx.get("scheme").asText());
    if (fx.has("publicKey")) assertThat(parsed.publicKey()).isEqualTo(fx.get("publicKey").asText());
    if (fx.has("host")) assertThat(parsed.host()).isEqualTo(fx.get("host").asText());
    if (fx.has("projectId")) assertThat(parsed.projectId()).isEqualTo(fx.get("projectId").asText());
    if (fx.has("ingestUrl")) assertThat(parsed.ingestUrl()).isEqualTo(fx.get("ingestUrl").asText());
  }
}
